bin/verdicts.py

```diff
@@ -148,11 +149,11 @@ class Verdicts:
 
     def __init__(
         self,
-        testcase_list: list[str],
+        testcases: list[testcase.Testcase],
         timeout: float = 1,
         run_until: RunUntil = RunUntil.FIRST_ERROR,
     ):
-        testcases = set(testcase_list)
+        testcases = {t.name for t in testcases}
```

Comment:
Ping

Reply:
ok, now i see it, but see reply below

Reply:
Ah, so this is where the disagreement lies.

First, just to get a misunderstanding out of the way, I prefer the `Verdict.__init__` parameter  to be an iterable over `str`. Formally, `testcases: Iterable[str]`. So I am _agnostic_ about `testcases: list[str]` versus `testcases: set[str]` versus any other kind of iterable. My point is about `testcases: Iterable[testcase.Testcase]` versus `testcases: Iterable[str]`.

Second, to me, a verdict is something that _a testcase_ (rather than a `testcase.Testcase`) has, and testcases are uniquely determined _by their name_ ,  such as `secret/tiny/02-bar` getting `AC`, maybe with `t=0.21`. 

Therefore I prefer `verdict.Verdict` to be _agnostic_ about _how_ a testcase is implemented (say, as a `testcase.Testcase` object in BAPCtools.) In extremis (and just to make the point), even `verifyproblem` should be able to `from verdict import Verdict` and then farm out all its verdicting needs.

Note that, for instance, `test_verdicts.py` imports only `import verdict` and that `verdicts.py` (at least in its original form) imports _almost nothing_ from the rest of BAPCtoools. (In particular, it imports neither `testcase` nor `run`.) 

Apart from just resulting from a desire to _reduce dependencies_ (in a very organic and complex codebase) it is also a result of me trying to define the concepts of **Verdict** (and the forthcoming **Expectation**) in a very general fashion, close to an attitude that mimicks the specification (and therefore shall not be tied to `testcase.Testcase`.)

In the end, these are opinions, and bleed into project-wide decisions about software architecture. I want to make my position clear, but am happy to be overruled.

Reply:
My point was more like:
- the call site for bapctools is nicer since it know takes as input the same list that we use at all other places where we want a list of testcases.
- the fact that we only need the name for `__init__` is an implementation detail. If at some point we want additional information about the testcase we can easily get it from `testcase.Testcase` but we can't easily get if from the name.
- regarding the dependencies: i think it would be really nice if we had some library part that just abstracts from the spec and encapsulates the directory layouts etc. and a bapctools part which contains the tooling... but right now its deeply tied together... ^^'
- from the programmer point of view: what is a testcase? Is it a name? is it the `.in`+`.ans` file? is it something else? I don't know but in the code its a `testcase.Testcase` that nicely abstracts whatever it is and gives me a simple interface to ask for.

Reply:
Yeah I'm with @mzuenni here; I think passing `Testcase` objects is more in line with the rest of BAPCtools.

I do agree that an independent type-agnostic library of parts of the spec may be nice, but that's not really in scope now. Anyway this is already strongly tied to BAPCtools I'd say since it also implements quite some logic to handle lazy juding, which is implied by the spec but not really part of it.

Reply:
You decide. (I disagree.)